### PR TITLE
Replace lazy_printv and printv_err with the built-in logging module

### PR DESCRIPTION
**Pull request by ilkilic**

### CODE REVIEW

**Comment by wvangeit** [comment]
Will this print the warnings etc without specific user intervention though? We need to make sure the users get the same warnings without them having to load the logging module etc.

**Comment by anilbey** [comment]
Thanks @ilkilic .


From the [docs](https://docs.python.org/3/howto/logging.html)

> The root logger always has an explicit level set (WARNING by default). 

@wvangeit the default setting for the verbosity level is WARNING. Nonetheless, it can be adjusted to various other levels such as DEBUG, INFO, or CRITICAL. 

At this point, I think we need to examine our log messages one by one to determine their appropriate labels, such as DEBUG, INFO, or WARNING.

For example, if we get a log message saying "an empty synapses data frame received," we should label it as WARNING. And if the message says "while the names of synaptic attributes are the following:", it's better to label it as DEBUG.

**Comment by ilkilic** [comment]
Indeed, it should print warnings without any specific user intervention. Additionally, users have the flexibility to adjust the verbosity level in tools.py (I chose INFO by default). So I tried to map the error levels as follows:

- WARNING: msg with an explicit "warning"
- CRITICAL:  msg with verbose level of 50
- ERROR: msg from printv_err
- INFO: everything else

I haven't used DEBUG I think, but let me know if I should change anything

**Review by anilbey — CHANGES_REQUESTED** [review]

File: bluecellulab/cell/core.py
Comment (ilkilic) [reply bluecellulab/cell/core.py:857]: Absolutely, thanks!

**Review by anilbey — APPROVED** [review]
Thanks looks very good.

 We are currently setting up a "dev" branch and a bot to automatically build and release bluecellulab. We will merge this PR once the repository's release bot is working fine.

**Comment by anilbey** [comment]
Rebased main.

File: bluecellulab/tools.py
Comment (wvangeit) [inline bluecellulab/tools.py:57]: Just to be sure. So if a user now call set_verbose(), the user won't have to do 'anything' else right to get the messages (i.e. no import of logging in his own scripts etc).

Comment (ilkilic) [reply bluecellulab/tools.py:57]: Correct, if a user calls set_verbose() it will configure the logging with a specific level. The user won't need to import the logging module or perform any additional setup in their own scripts to see the log messages.

**Review by wvangeit — APPROVED** [review]

File: bluecellulab/tools.py
Comment (anilbey) [reply bluecellulab/tools.py:57]: Actually it doesn't work on this case.

```python
Python 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from bluecellulab import RNGSettings
Setting rng mode to: Random123  # this gets printed as expected
>>> from bluecellulab.tools import set_verbose
>>> set_verbose(50)
>>> rng = RNGSettings()
Setting rng mode to: Random123  # this was not supposed to be printed
```

To solve it we probably need to set the level of root logger to affect all loggers at once.

```suggestion
    logging.getLogger().setLevel(level)  # set level of root logger
```

Comment (anilbey) [reply bluecellulab/tools.py:57]: I just noticed the docstring of that method says

> Verbose level, the higher the more verbosity.

Is it not the lower the more verbosity instead?

Comment (wvangeit) [reply bluecellulab/tools.py:57]: I thought 0 == nothing, 1 a bit, 100 all, etc?

**Review by wvangeit — CHANGES_REQUESTED** [review]
Blocking this for a sec until we're sure that logging behaves like we want :-)

File: bluecellulab/tools.py
Comment (anilbey) [reply bluecellulab/tools.py:57]: They have it the other way around. The higher, the less verbose.

https://docs.python.org/3/library/logging.html#logging-levels

Comment (anilbey) [reply bluecellulab/tools.py:57]: I didn't know that the logger names were defining a hierarchy by using dot-notation.
https://stackoverflow.com/a/50755200/1935611

In practice, our logger names are as follows

```python
(Pdb) __name__
'bluecellulab.cell.stimuli_generator'
```

To affect all of the bluecellulab loggers and (only the bluecellulab loggers alone) the following seems to be the solution. 

```python
logging.getLogger('bluecellulab').setLevel(level)
```

Having a unit test for ensuring it would be great.

Comment (wvangeit) [reply bluecellulab/tools.py:57]: ok, sure, sorry, you were talking about the logging module itself.

Comment (wvangeit) [reply bluecellulab/tools.py:57]: and yes, the hierarchy might be a solution indeed.

Comment (anilbey) [reply bluecellulab/tools.py:57]: > Yeah, but aren't we messing up the user settings with that? 

Yes right. We shouldn't touch the root logger. We should only touch the `bluecellulab.` and its children loggers.

Comment (ilkilic) [reply bluecellulab/tools.py:57]: I'll make the necessary changes then and also explain more clearly in the comments that the higher the level, the less verbose

Comment (wvangeit) [reply bluecellulab/tools.py:57]: We should not change the meaning of the argument of set_verbose API though. It should act in a similar way as before.

Comment (wvangeit) [reply bluecellulab/tools.py]: We should keep complete backward compatibility. If in the past somebody set set_verbose to something, the software should act as close as possible to the past behavior

Comment (wvangeit) [reply bluecellulab/tools.py]: Obviously, if the user interacts with the logging module directly, the behavior of the logging module itself doesn't (shouldn't) have to change.
But scripts written in the past (using bglibpy set_verbose e.g.) should still act the same as they did before.

File: bluecellulab/tools.py
Comment (wvangeit) [inline bluecellulab/tools.py:42]: Please also don't change the default level

**Review by wvangeit — APPROVED** [review]

**Review by anilbey — APPROVED** [review]
Brilliant, thanks @ilkilic

**Pull request merged**
